### DISCUSSION TITLE
Spring Libraries Development Best Practice

### DISCUSSION THREAD

**Discussion by stliu** [Microsoft/spring-cloud-azure#986] [Ideas]
Inspired by [this](http://java.jonathangiles.net/) and [this](https://jlbp.dev/)

1. When create a new class, make it in `implementation` package by default, also `non-public` by default, as stated [here](http://java.jonathangiles.net/JLBP-7)
2.

**Comment by chenrujun** [comment]
Bean definition best practice:

1. **XxxConfiguratoin** should put in **implementation** package. Because customer will not use these classes in their java code. We should expose public API as less as possible.
2. **XxxAutoConfiguration** should **NOT** put in **implementation** package. Because customer may use it like **@SpringBootApplication(exclude=XxxAutoConfiguration.class)**.
3. **spring.factories** only contains **XxxAutoConfiguration**. It should not contain **XxxConfiguration**. If **spring.factories** contains **XxxConfiguration**, customer cannot exclude **XxxConfiguration** in their application. Because customer should not use the classes in **implementation** package. And Java 9's modules system can hide **implementation** package. 
4. **XxxAutoConfiguration** should import **XxxConfiguratoin**s, should not define bean directly. The bean definition is implementation detail. It should not expose as public API.

Question:
1. If **XxxConfiguration** put in **implementation** package, and Java 9's modules system hide **implementation** package, Can Spring create bean according to **XxxConfiguration**?

**Reply by chenrujun** [reply]
Moved this into the private wiki inside Spring Cloud Azure team. Because it's not ready to publish.
